src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -316,7 +317,15 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     def asSelect(implicit ctx: Context): Tree = {
       val qual1 = typedExpr(tree.qualifier, selectionProto(tree.name, pt, this))
       if (tree.name.isTypeName) checkStable(qual1.tpe, qual1.pos)
-      typedSelect(tree, pt, qual1)
+      val select = typedSelect(tree, pt, qual1)
+      if (select.tpe != TryDynamicCallType) {
+        select
+      } else {
+        pt match {
+          case _: FunProto | _: AssignProto => select
+          case _                            => typedDynamicSelect(tree, pt)
+        }
```

Comment:
I'd refactor to

    pt match {
       case _: FunProto | _: AssignProto => select
       case _  => 
         if (select.tpe eq TryDynamicCallType) typedDynamicSelect(tree, pt)
         else select
    }